README.md
# Search and Replace v1.10

**Note**: since v1.9 '*Search and Replace*' use: [Golang GtkSourceView binding for use with gotk3](https://github.com/hfmrow/gotk3_gtksource)

#### Last update 2021-04-24

*This program is designed to find and replace a text pattern in one or more files stored in one or more directories, sub-directories.
Search and replace in the clipboard is allowed. Drag and drop can be used.
You can use it to search multiple files to find definitions of specific functions in a development environment for example.*

Take a look at [H.F.M repositories](https://github.com/hfmrow/) for others useful Linux software.

- If you just want **to use it**, simply **download** the *.deb package version under the [Releases](https://github.com/hfmrow/search-and-replace/releases) tab. 

- Otherwise, if you plan to play inside the source code, see below "How to compile" section.

- All suggestions, contributions and ideas to improve software usability will be greatly appreciated.

### How it's made

- Programmed with go language: [golang](https://golang.org/doc/)
- GUI provided by [Gotk3](https://github.com/gotk3/gotk3), GUI library for Go (minimum required gtk3.16).
- Text editor use [GtkSourceView](https://github.com/hfmrow/gotk3_gtksource) component.
- I use home-made software: "Gotk3ObjHandler" to embed images/icons, UI-information and manage/generate gtk3 objects code from [glade ui designer](https://glade.gnome.org/). and "Gotk3ObjTranslate" to generate the language files and the assignment of a tool-tip on the gtk3 objects (both are not published at the moment, in fact, they need documentations and, I have not had the time to do them).

### Functionalities

- Search and replace text based on pattern (contained in text file(s)).
- Search and replace text pattern into clipboard.
- A map that allows you to view the content of the text and allows easy navigation through it.
- Allow the displaying of preview with highlighted patterns before modification.
- Source code highlighting for multiple languages with navigation map.
- Wildcard, regex, case sensitive, character classes can be used.
- Whole word functionality, Escape character can be used.
- Drag and drop capacity.
- Files can be sorted by extension or using a mask for filename.
- Lines that contain the searched pattern can be independently selected to be modified or vice versa to keep the content unchanged.
- Options (top right) to filter text files analysis behavior (text detection) and to limit depth of directories where files will be searched.
- Backup function available.
- Each function have his tool-tip for explanations.

###### This is the main screen

![](assets/readme/main.jpg)

###### Found window

![Found window](assets/readme/found.jpg)

###### Preview window

![Preview window](assets/readme/preview.jpg)

### How to compile

- Be sure you have golang installed in right way. [Go installation](https://golang.org/doc/install).

- Open terminal window and at command prompt, type: `go get github.com/hfmrow/search-and-replace`

- Install [Go bindings for GTK3](https://github.com/gotk3/gotk3) and follow [Installation instructions](https://github.com/gotk3/gotk3/wiki#installation).

- Get [GtkSourceView](https://github.com/hfmrow/gotk3_gtksource) library and follow instructions to install it with his `libgtksourceview-X-dev` package. Required since version 1.9 see at the bottom.

- **Since Go 1.16, native embed library is used instead of previous one and the following restriction is out of date.**

- ~~To change gtk3 interface you need to set `devMode` flag at `true`. A home made software, (not published actually) have been used to generate some parts of source code / assets embedding. So, you cannot (at this time) change interface for production mode.~~

- To change language file you need to use another home made software, (not published actually). You can still do it manually, all data has been stored in a '.json' file in the `assets/lang` directory.

- To Produce a *stand-alone executable*, you must change inside `main.go` file:

```go
    func main() {
        devMode = true
    ...    
```

into

```go
    func main() {
        devMode = false
    ...
```

This operation indicates that the internal behavior of the software will be modified to adapt to the production environment (display of errors, location of the configuration file, etc.).

## Os information:

<details>
  <summary>Built using</summary>

| Name                                                       | Version / Info / Name                          |
| ---------------------------------------------------------- | ---------------------------------------------- |
| GOLANG                                                     | V1.16.3 -> GO111MODULE="off", GOPROXY="direct" |
| DISTRIB                                                    | LinuxMint Xfce                                 |
| VERSION                                                    | 20                                             |
| CODENAME                                                   | ulyana                                         |
| RELEASE                                                    | #46-Ubuntu SMP Fri Jul 10 00:24:02 UTC 2020    |
| UBUNTU_CODENAME                                            | focal                                          |
| KERNEL                                                     | 5.8.0-50-generic                               |
| HDWPLATFORM                                                | x86_64                                         |
| GTK+ 3                                                     | 3.24.20                                        |
| GLIB 2                                                     | 2.64.3                                         |
| CAIRO                                                      | 1.16.0                                         |
| [GtkSourceView](https://github.com/hfmrow/gotk3_gtksource) | 4.6.0                                          |
| [LiteIDE](https://github.com/visualfc/liteide)             | 37.4 qt5.x                                     |
| Qt5                                                        | 5.12.8 in /usr/lib/x86_64-linux-gnu            |

</details>

- The compilation have not been tested under Windows or Mac OS.

### You got an issue ?

- Go to this page: [Issues hfmrow/search-and-replace on GitHub](https://github.com/hfmrow/search-and-replace/issues) and start a new problem report.
- Give the information (as above), concerning your working environment as well as the version of the operating system used.
- Provide a method to reproduce the problem.

### Used libraries

- Since version 1.9: [Golang GtkSourceView binding for use with gotk3](https://github.com/hfmrow/gotk3_gtksource)
- ~~Before version 1.9: [A general purpose syntax highlighter in pure Go](https://github.com/alecthomas/chroma)~~
- [Go Humans! (formatters for units to human friendly time/date)](https://github.com/dustin/go-humanize)
- [Go bindings for GTK3](https://github.com/gotk3/gotk3)
- And some personal libraries not yet published.

### Visit

##### [GitHub repository](https://github.com/hfmrow/SearchAndReplace)

##### [Website H.F.M's Linux softwares](https://hfmrow.go.yo.fr/)


CHANGELOG.md
# Search and Replace

## information

At the bottom you can find a compiled standalone ".deb" version with its checksum. The ".tar.gz" sources contain a "vendor" directory ensuring you can always compile it even if the official libraries have been changed.

## Changelog

All notable changes to this project will be documented in this file.

#### [1.10] 2021-04-24

#### Added

- When the files are dropped, the parent directory is now considered as root, it is stored in the file selector, if the button "Scan files" is pressed , this directory will be fully analyzed.
- Adding expand all in found files window.
- Adding check boxes to un-select lines to be preserved during replace actions.
- Adding un-select all and invert selection in found window.

#### Fixed

- The text preview window does not correctly highlight the searched pattern when escaped characters [Esc search] are used. This is fixed and no longer occurs.

#### Changed

- The application can be closed during the search process, of course the operation will be interrupted.
- By clicking on the search button, if there is no selected file, it will be considered that all files are selected (previously, file selection was explicitly required). Of course, the explicit selection of file (s) remains possible.
- Some incompatible search combinations are now disabled depending on the options checked.

---

#### [1.9] 2021-04-02

Since this version: [Golang GtkSourceView binding for use with gotk3](https://github.com/hfmrow/gotk3_gtksource). `$ sudo apt install libgtksourceview-4-dev` is required to be compiled.

If your OS does not implement GtkSourceView library natively, you can install it using:

`$ sudo apt install libgtksourceview-4-0`

If your Debian version is < **focal** (*20.04*), this means like **xenial** (*16.04*), **bionic** (*18.04*), this version does not work, it must re-compiled with appropriate `libgtksourceview-3.0-dev` library and version restrictions commands:

`$ go build -tags "gtksourceview_3_24" source.go`

#### Added

- Improving Source-code displaying [500ms for source > 350Kb]
- Allow the use of 'escape' characters with replace pattern. An option was added to enable or not the use of it.
- Source map is now included in preview window.

#### Fixed

- Application close while adding numbers to text > 360Kb, fixed using [GtkSourceView](https://github.com/hfmrow/gotk3_gtksource)

#### Changed

- Repository name was changed to [https://github.com/hfmrow/search-and-replace](https://github.com/hfmrow/search-and-replace) instead of `https://github.com/hfmrow/sAndReplace`
- A lot of source code has been modified to be able to use the features described below.
- Text numbering provided by [GtkSourceView](https://github.com/hfmrow/gotk3_gtksource) now.
- Preview search is now provided using [GtkSourceView](https://github.com/hfmrow/gotk3_gtksource) library
- Syntax highlighting is now provided by [GtkSourceView](https://github.com/hfmrow/gotk3_gtksource) too
- Code updated to fit actual version of [GitHub - Go bindings for GTK3](https://github.com/gotk3/gotk3)

---

#### [1.8] 2020-08-06

#### Added

- Text preview window: context menu allow you to open the directory that contain the current viewed file or opening it using your default file viewer. It's configured to work with most of desktop environement "*xdg-open*", you may change that inside ".opt" file to fit your's if that doesn't work.

- Files list main window: context menu allow you to open selected file or directory that contain the current selection. It's configured to work with most of desktop environement "*xdg-open*", you may change that inside ".opt" file to fit your's if that doesn't work.

- Syntax highlighting cover multiple languages and styles.

- Options window added (icon at top right).

- Option to ignore files where size does not match lower and higher values defined as byte.

#### Fixed

- Some little issues using preview window while lines number are visible.

- Search using escape characters match correctly when pattern contain regex' reserved characters [*.].

#### Changed

- Switch-control to enabling/disabling file-chooser-button has been changed to control the auto refresh on directory change.

- Some options was moved to new options window.

- Global artwork changed, icons, top image.

- The search notifications (information at the top right) using button, have been replaced by a progress bar.

- Found files window and preview window now keep there place on screen when reopened in the same session.

- Preview window: the function that allow line number display was rewritten. Better handling large text. More precision when placing cursor at row number containing the found pattern, and better global stability.

- Cleaning some useless code.

- Rewriting some others part of code to improve speed while searching in files.

- Code updated to fit actual version of [GitHub - Go bindings for GTK3](https://github.com/gotk3/gotk3)

---

#### [1.7.6] 2019-10-02

#### Added

- Adding lines numbers in preview window.

- Highlight found patterns in preview window.

- Option to show/hide forbidden files in result window, (binary and those < 8 bytes).

- Adding filenames, found count in title of the preview window and in found result window.

- The found result windows, now allow to double-click on sub nodes to show it in preview window and positioning preview to specified line.

#### Fixed

- Some issues concerning directories scanning, Dnd and/or command line handling. Now it  preserve the chosen source files (filechooser act independently) a switch prevent unwanted changes.

#### Changed

- Scanning sub directory now handle depth of directory scanning.

- Lot of parts of the code have been rewritten for more stability and a really must faster processing.

- The interface has been reorganized.

---

#### [1.7.3] 2019-07-19

#### Added

- Switch-control to enabling/disabling file-chooser-button on command line argument passing.

#### Fixed


gohOptions.go
// gohOptions.go

/*
	Source file auto-generated on Sat, 24 Apr 2021 04:33:11 using Gotk3 Objects Handler v1.7.8
	©2018-21 hfmrow https://hfmrow.github.io

	This software use gotk3 that is licensed under the ISC License:
	https://github.com/gotk3/gotk3/blob/master/LICENSE

	Copyright ©2018-21 hfmrow - Search And Replace v1.10 github.com/hfmrow/search-and-replace

	This program comes with absolutely no warranty. See the The MIT License (MIT) for details:
	https://opensource.org/licenses/mit-license.php
*/

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"

	glfs "github.com/hfmrow/genLib/files"
	glfsft "github.com/hfmrow/genLib/files/fileText"
	glfssf "github.com/hfmrow/genLib/files/scanFileDir"
	glss "github.com/hfmrow/genLib/slices"
	glsg "github.com/hfmrow/genLib/strings"
	glts "github.com/hfmrow/genLib/tools"
	gltsbh "github.com/hfmrow/genLib/tools/bench"
	gltsle "github.com/hfmrow/genLib/tools/log2file"
	gltsushe "github.com/hfmrow/genLib/tools/units/human_readable"

	gidg "github.com/hfmrow/gtk3Import/dialog"
	gimc "github.com/hfmrow/gtk3Import/misc"
	gipops "github.com/hfmrow/gtk3Import/pango/pangoSimple"
	gitvsv "github.com/hfmrow/gtk3Import/textView/sourceView"
	gitl "github.com/hfmrow/gtk3Import/tools"
	gits "github.com/hfmrow/gtk3Import/tools"
	gitw "github.com/hfmrow/gtk3Import/treeview"
)

// App infos
var (
	Name         = "Search And Replace"
	Vers         = "v1.10"
	Descr        = "Search and replace text sequences in one or more files.\nThe clipboard is also included inside replacing functionality.\nThere is some useful functionalities like code preview with\nsyntax highlighting, backup modified files whether is requested,\nand many more tricks."
	Creat        = "hfmrow"
	YearCreat    = "2018-21"
	LicenseShort = "This program comes with absolutely no warranty.\nSee the The MIT License (MIT) for details:\nhttps://opensource.org/licenses/mit-license.php"
	LicenseAbrv  = "License (MIT)"
	Repository   = "github.com/hfmrow/search-and-replace"
)

// Vars declarations
var (
	absoluteRealPath,
	optFilename string

	opt *MainOpt
	tempDir,
	currFilename,
	mainWindowTitle string

	doTempDir,
	namingWidget,
	devMode,
	acceptBinary bool
	filesCount,
	currentLine int
	textWinTextToShowBytes []byte

	/*
	 * Lib mapping
	 */
	// AboutBox
	AboutInfosNew = gidg.AboutInfosNew
	About         *gidg.AboutInfos

	tvsTree,
	tvsList *gitw.TreeViewStructure
	TreeViewStructureNew = gitw.TreeViewStructureNew

	statusbar                                *gimc.StatusBar
	mainWinTitle, textWinTitle, findWinTitle *gimc.TitleBar
	dnd                                      *gimc.DragNDropStruct
	DragNDropNew                             = gimc.DragNDropNew
	TitleBarStructureNew                     = gimc.TitleBarStructureNew
	StatusBarStructureNew                    = gimc.StatusBarStructureNew

	// Popup menu
	PopupMenuIconStructNew = gimc.PopupMenuIconStructNew
	popMenuTextView        *gimc.PopupMenuIconStruct
	popupMenu              *gimc.PopupMenuIconStruct

	DialogMessage = gidg.DialogMessage

	GetEntryText      = gitl.GetEntryText
	GetEntryTextAsInt = gitl.GetEntryTextAsInt

	pbs            *gimc.ProgressBarStruct
	ProgressGifNew = gimc.ProgressGifNew

	PangoColorNew = gipops.PangoColorNew

	// Error handling
	DlgErr = func(dsc string, err error) (yes bool) {
		yes = gidg.DialogError(obj.mainWin, sts["issue"], dsc, err, devMode, devMode)
		return
	}
	// Errors handling
	Log2FileStructNew = gltsle.Log2FileStructNew
	Logger            *gltsle.Log2FileStruct

	// GtkSourceView
	svs                 *gitvsv.SourceViewStruct
	SourceViewStructNew = gitvsv.SourceViewStructNew

	// Genlib decl
	filesFoundMulti         []glfsft.SearchAndReplaceFiles
	fileFoundSingle         *glfsft.SearchAndReplace
	SearchAndReplaceNew     = glfsft.SearchAndReplaceNew
	SearchAndReplaceInFiles = glfsft.SearchAndReplaceInFiles
	IsTextFile              = glfsft.IsTextFile

	IsExistSl         = glss.IsExistSl
	IsExistSlIface    = glss.IsExistSlIface
	DeleteSlIface     = glss.DeleteSlIface
	HumanReadableSize = gltsushe.HumanReadableSize
	TruncatePath      = glfs.TruncatePath
	RemoveNonNum      = glsg.RemoveNonNum
	UnEscapedStr      = func(in string) string {
		if obj.chkUseEscapeChar.GetActive() {
			return glsg.UnEscapedStr(in)
		}
		return in
	}
	// UnEscapeString = glsg.UnEscapeString

	BenchNew    = gltsbh.BenchNew
	ExecCommand = glts.ExecCommand

	// Files
	toDispFileList       []glfssf.ScanDirFileInfos
	ScanDirDepthFileInfo = glfssf.ScanDirDepthFileInfo

	// Global vars
	previsouWordsPos   [][]int
	currentInFilesList []string
	scanTime,
	searchTime,
	currentText,
	dispTime string // Some times controls

	currentTextChanged,
	forbiddenFiles,
	forbiddenFilesAlreadyDone,
	fromDnD,
	alreadyPlacedPrevWin,
	alreadyPlacedFoundWin,
	btnFindInUse,
	fileChooserDoUpdt,
	displayFilesReadyToUpdate bool
)

type MainOpt struct {
	// File signature
	FileSign []string

	MainWinWidth,
	MainWinHeight,
	MainWinPosX,
	MainWinPosY,
	PanedWidth,
	SourceWinWidth,
	SourceWinHeight,
	SourceWinPosX,
	SourceWinPosY int

	FixedMapWidth bool

	LanguageFilename string

	CaseSensitive,
	UseRegex,
	WholeWord,
	ScanSubDir,
	FollowSymlinkDir,
	MakeBackup,
	UseCharacterClass,
	UseWildcard,
	WrapText,
	DispForbiddenFiles,
	ExpandAll,
	UseEscapeChar,
	UseEscapeCharToReplace,
	AutoScan,
	SyntaxHighlight bool

	ExtMask []string

	ExtSep,
	Directory,
	AppLauncher,
	DefaultSourceLang,
	DefaultSourceStyle,
	HighlightUserDefined,
	TxtBgCol,
	TxtFgCol string

	CascadeDepth,
	FilePathLength,
	ScanDirDepth int

	FileMinSizeLimit,
	FileMaxSizeLimit int64

	/* Private, will NOT be saved */
	listStoreColumns,
	treeStoreColumns [][]string
	mapListStore,
	mapTreeStore map[string]int
}

// Main options initialisation
func (opt *MainOpt) Init() {

	opt.MainWinWidth = 800
	opt.MainWinHeight = 600
	opt.CascadeDepth = 30
	opt.FilePathLength = 4

	opt.PanedWidth = 120
	opt.SourceWinWidth = 800
	opt.SourceWinHeight = 480
	opt.SourceWinPosX = -1
	opt.SourceWinPosY = -1

	opt.LanguageFilename = "assets/lang/eng.lang"
	opt.Directory = ""
	opt.ExtMask = []string{"*"}
	opt.ExtSep = ";"
	opt.listStoreColumns = [][]string{
		{"Name", "text"},
		{"Size", "text"},
		{"Date", "text"},
		{"Path", "markup"},
		{"pathReal", "text"},  // This one will be invisible (int64)
		{"sizeSort", "int64"}, // This one will be invisible (int64)
		{"dateSort", "int64"}} // This one will be invisible (int64)}
	opt.mapListStore = map[string]int{
		"Name":     0,
		"Size":     1,
		"Date":     2,
		"PathDisp": 3,
		"pathReal": 4,
		"sizeSort": 5,
		"dateSort": 6}
	opt.treeStoreColumns = [][]string{
		{"", "active"},
		{"Filename(s)", "markup"},
		{"fileIdx", "int64"},
		{"lineIdx", "int64"}}
	opt.mapTreeStore = map[string]int{
		"Toggle":   0,
		"Filename": 1,
		"fileIdx":  2, // This one will be invisible (int64)
		"lineIdx":  3} // This one will be invisible (int64)

	opt.ScanDirDepth = -1
	opt.AutoScan = true
	opt.MakeBackup = true

	opt.AppLauncher = "xdg-open"

	opt.DefaultSourceLang = "go-hfmrow"
	opt.DefaultSourceStyle = "hfmrow"
	opt.HighlightUserDefined = "assets/langAndstyle"
	opt.SyntaxHighlight = true

	opt.TxtBgCol = "#F8F8F8"
	opt.TxtFgCol = "#1A1A1A"

	opt.FileMinSizeLimit = 16      // means 16 bytes
	opt.FileMaxSizeLimit = 1048576 // means 1M/b
}

// Variables -> Objects.
func (opt *MainOpt) UpdateObjects() {
	obj.mainWin.Resize(opt.MainWinWidth, opt.MainWinHeight)
	obj.mainWin.Move(opt.MainWinPosX, opt.MainWinPosY)

	obj.textWin.Resize(opt.SourceWinWidth, opt.SourceWinHeight)
	obj.textWin.Move(opt.SourceWinPosX, opt.SourceWinPosY)
	obj.Paned.SetPosition(opt.SourceWinWidth - opt.PanedWidth)

	OptToExtSlice()
	obj.fileChooserBtn.SetCurrentFolder(opt.Directory)
	obj.chkCaseSensitive.SetActive(opt.CaseSensitive)
	obj.chkRegex.SetActive(opt.UseRegex)
	obj.chkWholeWord.SetActive(opt.WholeWord)
	obj.chkWildcard.SetActive(opt.UseWildcard)
	obj.chkCharacterClass.SetActive(opt.UseCharacterClass)
	obj.textWinChkWrap.SetActive(opt.WrapText)
	obj.findWinChkBackUp.SetActive(opt.MakeBackup)
	obj.findWinChkDispForbFiles.SetActive(opt.DispForbiddenFiles)
	obj.chkFollowSymlinkDir.SetActive(opt.FollowSymlinkDir)
	obj.textWinChkSyntxHighlight.SetActive(opt.SyntaxHighlight)

	obj.chkUseEscapeChar.SetActive(opt.UseEscapeChar)
	obj.chkUseEscapeCharToReplace.SetActive(opt.UseEscapeCharToReplace)

	obj.OptionsEntryMaxFileSize.SetText(fmt.Sprintf("%d", opt.FileMaxSizeLimit))
	obj.OptionsEntryMinFileSize.SetText(fmt.Sprintf("%d", opt.FileMinSizeLimit))
	obj.switchFileChooserButton.SetActive(opt.AutoScan)
	obj.SourceToggleButtonMapWidth.SetActive(opt.FixedMapWidth)

	obj.findWinChkExpandAll.SetActive(opt.ExpandAll)

	/* Init spinButton */
	gits.SpinScaleSetNew(obj.spinButtonDepth, -1, 128, float64(opt.ScanDirDepth), 1)
}

// Objects -> Variables.
func (opt *MainOpt) UpdateOptions() {
	opt.MainWinWidth, opt.MainWinHeight = obj.mainWin.GetSize()
	opt.MainWinPosX, opt.MainWinPosY = obj.mainWin.GetPosition()

	opt.SourceWinWidth, opt.SourceWinHeight = obj.textWin.GetSize()
	opt.SourceWinPosX, opt.SourceWinPosY = obj.textWin.GetPosition()
	opt.PanedWidth = opt.SourceWinWidth - obj.Paned.GetPosition()
	opt.FixedMapWidth = obj.SourceToggleButtonMapWidth.GetActive()

	ExtSliceToOpt() // Store extensions list
	opt.Directory = obj.fileChooserBtn.FileChooser.GetFilename()
	opt.CaseSensitive = obj.chkCaseSensitive.GetActive()
	opt.UseRegex = obj.chkRegex.GetActive()
	opt.WholeWord = obj.chkWholeWord.GetActive()
	opt.UseWildcard = obj.chkWildcard.GetActive()
	opt.UseCharacterClass = obj.chkCharacterClass.GetActive()
	opt.WrapText = obj.textWinChkWrap.GetActive()
	opt.MakeBackup = obj.findWinChkBackUp.GetActive()
	opt.DispForbiddenFiles = obj.findWinChkDispForbFiles.GetActive()
	opt.FollowSymlinkDir = obj.chkFollowSymlinkDir.GetActive()
	opt.SyntaxHighlight = obj.textWinChkSyntxHighlight.GetActive()

	opt.UseEscapeChar = obj.chkUseEscapeChar.GetActive()
	opt.UseEscapeCharToReplace = obj.chkUseEscapeCharToReplace.GetActive()

	// Value are sanitized, so, we only need to retrieve them
	opt.FileMaxSizeLimit = int64(gits.GetEntryTextAsInt(obj.OptionsEntryMaxFileSize))
	opt.FileMinSizeLimit = int64(gits.GetEntryTextAsInt(obj.OptionsEntryMinFileSize))
	opt.AutoScan = obj.switchFileChooserButton.GetActive()

	opt.ExpandAll = obj.findWinChkExpandAll.GetActive()

	opt.ScanDirDepth = obj.spinButtonDepth.GetValueAsInt()
}

// Read Options from file
func (opt *MainOpt) Read() (err error) {
	var textFileBytes []byte
	opt.Init() // Init with default values.
	if textFileBytes, err = ioutil.ReadFile(optFilename); err == nil {
		err = json.Unmarshal(textFileBytes, &opt)
	}
	return err
}

// Write Options to file
func (opt *MainOpt) Write() (err error) {
	var out bytes.Buffer
	var jsonData []byte
	opt.UpdateOptions()
	opt.FileSign = []string{Name, Vers, "©" + YearCreat, Creat, Repository, LicenseAbrv}
	if jsonData, err = json.Marshal(&opt); err == nil {
		if err = json.Indent(&out, jsonData, "", "\t"); err == nil {
			err = ioutil.WriteFile(optFilename, out.Bytes(), os.ModePerm)
		}
	}
	return err
}
